arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/parTupledN/ParTupled4Test.kt

```diff
@@ -0,0 +1,195 @@
+        }
+      )
+
+      r.get() shouldBe "$d$c$b$a"
+    }
+  }
+
+  "ParTupledN 4 finishes on single thread" {
+    checkAll(Arb.string()) { name ->
+      single.use { ctx ->
+        parTupledN(ctx, threadName, threadName, threadName, threadName)
+      } shouldBe Tuple4("single", "single", "single", "single")
+    }
+  }
+
+  "Cancelling ParTupledN 4 cancels all participants" {
+    checkAll(Arb.int(), Arb.int(), Arb.int(), Arb.int()) { a, b, c, d ->
+      val s = Semaphore(0L)
+      val pa = Promise<Pair<Int, ExitCase>>()
+      val pb = Promise<Pair<Int, ExitCase>>()
+      val pc = Promise<Pair<Int, ExitCase>>()
+      val pd = Promise<Pair<Int, ExitCase>>()
+
+      val loserA = suspend { guaranteeCase({ s.release(); never<Int>() }) { ex -> pa.complete(Pair(a, ex)) } }
+      val loserB = suspend { guaranteeCase({ s.release(); never<Int>() }) { ex -> pb.complete(Pair(b, ex)) } }
+      val loserC = suspend { guaranteeCase({ s.release(); never<Int>() }) { ex -> pc.complete(Pair(c, ex)) } }
+      val loserD = suspend { guaranteeCase({ s.release(); never<Int>() }) { ex -> pd.complete(Pair(d, ex)) } }
+
+      val f = ForkAndForget { parTupledN(loserA, loserB, loserC, loserD) }
+
+      s.acquireN(4) // Suspend until all racers started
+      f.cancel()
+
+      pa.get().let { (res, exit) ->
+        res shouldBe a
+        exit.shouldBeInstanceOf<ExitCase.Cancelled>()
+      }
+      pb.get().let { (res, exit) ->
+        res shouldBe b
+        exit.shouldBeInstanceOf<ExitCase.Cancelled>()
+      }
+      pc.get().let { (res, exit) ->
+        res shouldBe c
+        exit.shouldBeInstanceOf<ExitCase.Cancelled>()
+      }
+      pd.get().let { (res, exit) ->
+        res shouldBe d
+        exit.shouldBeInstanceOf<ExitCase.Cancelled>()
+      }
+    }
+  }
+
+  "ParTupledN 4 cancels losers if a failure occurs in one of the tasks" {
+    checkAll(
+      Arb.throwable(),
+      Arb.element(listOf(1, 2, 3, 4)),
+      Arb.int(),
+      Arb.int(),
+      Arb.int()
+    ) { e, leftWinner, a, b, c ->
+      val s = Semaphore(0L)
+      val pa = Promise<Pair<Int, ExitCase>>()
+      val pb = Promise<Pair<Int, ExitCase>>()
+      val pc = Promise<Pair<Int, ExitCase>>()
+
+      val winner = suspend { s.acquire(); throw e }
+      val loserA = suspend { guaranteeCase({ s.release(); never<Int>() }) { ex -> pa.complete(Pair(a, ex)) } }
+      val loserB = suspend { guaranteeCase({ s.release(); never<Int>() }) { ex -> pb.complete(Pair(b, ex)) } }
+      val loserC = suspend { guaranteeCase({ s.release(); never<Int>() }) { ex -> pc.complete(Pair(c, ex)) } }
+
+      val r = Either.catch {
+        when (leftWinner) {
+          1 -> parTupledN(winner, loserA, loserB, loserC)
+          2 -> parTupledN(loserA, winner, loserB, loserC)
+          3 -> parTupledN(loserA, loserB, winner, loserC)
+          else -> parTupledN(loserA, loserB, loserC, winner)
+        }
+      }
+
+      pa.get().let { (res, exit) ->
```

Comment:
This check is missing for pb and pc. 🙈

Reply:
These tests are not super important anymore, and we should think about how we want to improve them. `parMapN` is now derived from `coroutineScope` and `async`, so all these scenarios should already be tested in the KotlinX side.
`parTupledN` is derived from `parMapN` which also already covers all these cases...

So we are running a lot of redundant tests, however I think many of these tests still make sense to make sure we derive the code from KotlinX correctly...

Reply:
Okay then, I'll leave it as it is.